PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc

```diff
@@ -815,6 +831,82 @@ double PATMuonProducer::getRelMiniIsoPUCorrected(const pat::Muon& muon, float rh
   return pat::muonRelMiniIsoPUCorrected(muon.miniPFIsolation(), muon.p4(), drcut, rho);
 }
 
+
+double PATMuonProducer::puppiCombinedIsolation(const pat::Muon& muon, const pat::PackedCandidateCollection *pc)
+{
+  double dR_threshold = 0.4;
+  double dR2_threshold = dR_threshold * dR_threshold;
+  double mix_fraction = 0.5;
+  enum particleType{
+      CH = 0,
+      NH = 1,
+      PH = 2,
+      OTHER = 100000
+    };
+  double val_PuppiWithLep = 0.0;
+  double val_PuppiWithoutLep = 0.0;  
+
+  for(const auto & cand : *pc){//pat::pat::PackedCandidate loop start
+
+     const particleType pType =
+        isChargedHadron( cand.pdgId() ) ? CH :
+        isNeutralHadron( cand.pdgId() ) ? NH :
+        isPhoton( cand.pdgId() ) ? PH : OTHER;
+     if( pType == OTHER ){
+        if( cand.pdgId() != 1 && cand.pdgId() != 2
+        && abs( cand.pdgId() ) != 11
+        && abs( cand.pdgId() ) != 13){
+        LogTrace("PATMuonProducer") <<"candidate with PDGID = " << cand.pdgId() << " is not CH/NH/PH/e/mu or 1/2 (and this is removed from isolation calculation)" << std::endl;
+       }
+       continue;
+     }
+     double d_eta = std::abs( cand.eta() - muon.eta() );
+     if( d_eta > dR_threshold ) continue;
+
+     double d_phi = std::abs(reco::deltaPhi(cand.phi(),muon.phi()));
+     if( d_phi > dR_threshold ) continue ;
+
+     double dR2=reco::deltaR2(cand, muon);
+     if( dR2  > dR2_threshold ) continue;
+     if( pType == CH && dR2 < 0.0001*0.0001 ) continue;
+     if( pType == NH && dR2 < 0.01  *0.01   ) continue;
+     if( pType == PH && dR2 < 0.01  *0.01   ) continue;
+     val_PuppiWithLep += cand.pt() * cand.puppiWeight();
+     val_PuppiWithoutLep += cand.pt() * cand.puppiWeightNoLep();
+
+  }//pat::pat::PackedCandidate loop end
+
+  double reliso_Puppi_withLep    = val_PuppiWithLep/muon.pt();
+  double reliso_Puppi_withoutlep = val_PuppiWithoutLep/muon.pt();
+  double reliso_Puppi_combined = mix_fraction * reliso_Puppi_withLep + ( 1.0 - mix_fraction) * reliso_Puppi_withoutlep;
+  return reliso_Puppi_combined;
+}
+
+bool PATMuonProducer::isNeutralHadron( long pdgid ){
+  const long id = abs( pdgid );
+  if( id == 111 ) return true ;
+  if( id == 130 ) return true ;
+  if( id == 310 ) return true ;
+  if( id == 2112 ) return true ;
+  return false;
+
+}
+
+bool PATMuonProducer::isChargedHadron( long pdgid ){
+  const long id = abs( pdgid );
+  if( id == 211    ) return true ;
+  if( id == 321    ) return true ;
+  if( id == 999211 ) return true ;
+  if( id == 2212   ) return true ;
+  return false;
```

Comment:
we are dealing with pf/packed candidates and there is only one ID per CH.
replace the body with ```return std::abs(pdgid) == 211;```